### PR TITLE
Feature/ocpp security extensions

### LINKED ISSUE #90: OCPP 1.6-J Security
The OCPP 1.6 standard has just changed. See:

[http://www.v2g-evse.com/2018/12/18/open-charge-alliance-enhances-ocpp-1-6-security/ ](http://www.v2g-evse.com/2018/12/18/open-charge-alliance-enhances-ocpp-1-6-security/ )

> Secure communication and operation is a critical aspect of Electric Vehicle Charging Infrastructure.
> 
> To further assist the industry the Open Charge Alliance now publishes a white paper to describe a standard way to address security using OCPP 1.6-J.

Jim

### PR DESCRIPTION
**Pull request by m-oben**
Resolves #90

### CODE REVIEW

File: ocpp-v1_6/src/test/java/eu/chargetime/ocpp/model/securityext/test/types/LogParametersTypeTest.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    LogParametersType type = new LogParametersType();
+
+    // When
+    type.setRemoteLocation(location);
+
+    // Then throws
+  }
+
+  private String givenLocation() {
+    return "Location";
```
Comment (TVolden) [inline ocpp-v1_6/src/test/java/eu/chargetime/ocpp/model/securityext/test/types/LogParametersTypeTest.java:100]: Should probably be an URL like "http://someone@somewhere.whatever/file.log?query=test&test=query"

**Review by TVolden — CHANGES_REQUESTED** [review]
This was an impressive change set. I have tried to compare the code to the white paper specification from OCA's homepage. I may have missed something, but I did my best.
Some general notes: I'm very thankful that you decided to add unit tests. I would have loved to have some behavior tests with some fake client/servers actually sending request/confirmations. I don't know if the deserializer can actually instantiate the requests/confirmations without an empty default constructor? Have you tested this?
Also, you seem to make boundary tests, but only test the negative scenarios, I think I found a bug that would have been caught if you also made positive scenarios (testing the very limits). 

Overall a very impressive change set, thank you for that. I only have some specification breaking issues I need you to fix before I can approve. Please see the comments for more information and please don't hesitate to ask if I'm writing jiberish again :)

File: ocpp-v1_6/src/main/java/eu/chargetime/ocpp/model/securityext/SecurityEventNotificationRequest.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+  private static final transient Validator typeValidator =
+    new ValidatorBuilder()
+      .addRule(new StringMaxLengthValidationRule(50))
+      .setRequired(true)
+      .build();
+
+  private static final transient Validator techInfoValidator =
+    new ValidatorBuilder()
+      .addRule(new StringMaxLengthValidationRule(255))
```
[the remark this answers is not in the document]
Comment (m-oben) [reply ocpp-v1_6/src/main/java/eu/chargetime/ocpp/model/securityext/SecurityEventNotificationRequest.java:48]: Indeed, it should be optional as the cardinality of the "techInfo" field is defined as 0..1.

File: ocpp-v1_6/src/main/java/eu/chargetime/ocpp/model/securityext/types/FirmwareType.java [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    return signingCertificate;
+  }
+
+  /**
+   * Required. Certificate with which the firmware was signed. PEM encoded X.509 certificate.
+   *
+   * @param signingCertificate string[0..5500]
+   */
+  public void setSigningCertificate(String signingCertificate) {
+    signatureValidator.validate(signingCertificate);
```
[the remark this answers is not in the document]
Comment (m-oben) [reply ocpp-v1_6/src/main/java/eu/chargetime/ocpp/model/securityext/types/FirmwareType.java:139]: Correct, I will fix it.

File: ocpp-v1_6/src/main/java/eu/chargetime/ocpp/model/securityext/types/LogParametersType.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+import eu.chargetime.ocpp.model.validation.ValidatorBuilder;
+import eu.chargetime.ocpp.utilities.MoreObjects;
+
+import java.time.ZonedDateTime;
+import java.util.Objects;
+
+public class LogParametersType implements Validatable {
+  private static final transient Validator remoteLocationValidator =
+    new ValidatorBuilder()
+      .addRule(OCPPSecurityExtDatatypes.identifierString())
```
[the remark this answers is not in the document]
Comment (m-oben) [reply ocpp-v1_6/src/main/java/eu/chargetime/ocpp/model/securityext/types/LogParametersType.java:41]: Correct, I will fix it.

File: ocpp-v1_6/src/main/java/eu/chargetime/ocpp/model/securityext/InstallCertificateRequest.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+  private CertificateUseEnumType certificateType;
+  private String certificate;
+
+  /**
+   * Handle required fields.
+   *
+   * @param certificateType CertificateUseEnumType. See {@link #setCertificateType(CertificateUseEnumType)}
+   * @param certificate     String. See {@link #setCertificate(String)}
+   */
+  public InstallCertificateRequest(CertificateUseEnumType certificateType, String certificate) {
```
[the remark this answers is not in the document]
Comment (m-oben) [reply ocpp-v1_6/src/main/java/eu/chargetime/ocpp/model/securityext/InstallCertificateRequest.java:55]: Good point, indeed I have written no tests to check for correct serialization. Most likely there will be issues. I'll check it out. This is the case for every model I've added.

File: ocpp-v1_6/src/main/java/eu/chargetime/ocpp/model/validation/IdentifierStringValidationRule.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+   SOFTWARE.
+*/
+
+import eu.chargetime.ocpp.PropertyConstraintException;
+
+import java.util.regex.Pattern;
+
+public class IdentifierStringValidationRule implements IValidationRule {
+  private final String ERROR_MESSAGE = "Illegal character(s) in IdentifierString.";
+  private final String PATTERN = "([a-zA-Z0-9]|\\*|\\-|\\_|\\=|\\:|\\+|\\||\\@|\\.)+";
```
[the remark this answers is not in the document]
Comment (m-oben) [reply ocpp-v1_6/src/main/java/eu/chargetime/ocpp/model/validation/IdentifierStringValidationRule.java:34]: IdentifierStringValidationRule is actually a copy from the ocpp 2.0 part of the repo. Reason it is a copy is that the whitepaper itself is a backport from ocpp 2.0 to 1.6. You wrote this once ;-) See here: ocpp-v2_0/src/main/java/eu/chargetime/ocpp/model/validation/IdentifierStringValidationRule.java

File: ocpp-v1_6/src/test/java/eu/chargetime/ocpp/model/securityext/test/types/LogParametersTypeTest.java (hunk 1 above)
Comment (m-oben) [reply ocpp-v1_6/src/test/java/eu/chargetime/ocpp/model/securityext/test/types/LogParametersTypeTest.java:100]: You are correct, however as you can see in most of my test cases (also for certificate strings etc), I just use arbitrary strings (or often identifier strings) to make sure the class under test performs as expected. That is, a string entered through the constructor/setter needs needs to be valid according to the spec. Your location url example is as valid as simply "Location". However, on application level, of course your example is better. But according to the spec, in theory any string can be entered.

File: ocpp-v1_6/src/main/java/eu/chargetime/ocpp/model/validation/IdentifierStringValidationRule.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (TVolden) [reply ocpp-v1_6/src/main/java/eu/chargetime/ocpp/model/validation/IdentifierStringValidationRule.java:34]: Yes of course, thank you. I was looking in the 1.6 specification.

File: ocpp-v1_6/src/main/java/eu/chargetime/ocpp/model/securityext/InstallCertificateRequest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (TVolden) [reply ocpp-v1_6/src/main/java/eu/chargetime/ocpp/model/securityext/InstallCertificateRequest.java:55]: If the serializer can work without a default empty constructor then I would prefer to have none. To test it I would make a simple fake client and server implementation to send an actual message via the system. This was what I did originally using the spock framework.

[the remark this answers is not in the document]
Comment (TVolden) [reply ocpp-v1_6/src/main/java/eu/chargetime/ocpp/model/securityext/InstallCertificateRequest.java:55]: Uh exciting, hope you like the framework 🙂. I see your point, private constructors sounds good.

File: ocpp-v1_6/src/test/java/eu/chargetime/ocpp/model/securityext/test/SecurityEventNotificationRequestTest.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    ZonedDateTime timestamp = ZonedDateTime.now();
+
+    // When
+    new SecurityEventNotificationRequest(type, timestamp);
+
+    // Then throws
+  }
+
+  @Test(expected = PropertyConstraintException.class)
+  public void setTechInfo_exceedingLengthString_throwsPropertyConstraintException() {
```
[the remark this answers is not in the document]
Comment (m-oben) [reply ocpp-v1_6/src/test/java/eu/chargetime/ocpp/model/securityext/test/SecurityEventNotificationRequestTest.java:118]: Added test cases for all models that have maximum length strings.

File: ocpp-v1_6/src/test/java/eu/chargetime/ocpp/feature/profile/test/securityext/ClientSecurityExtProfileTest.java [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+  @Mock
+  private ClientSecurityExtEventHandler handler;
+
+  @Before
+  public void setup() {
+    profile = new ClientSecurityExtProfile(handler);
+  }
+
+  @Test
+  public void getFeatureList_containsAllFeature() {
```
[the remark this answers is not in the document]
Comment (m-oben) [reply ocpp-v1_6/src/test/java/eu/chargetime/ocpp/feature/profile/test/securityext/ClientSecurityExtProfileTest.java:95]: The idea is that I test against the state of an object given a scenario/state. In this case it is the default state (no "given"), and as a result I expect all those features to be present in the list. As you said multiple tests in theory do the same thing (+ added value of possible multiple features that are missing), but it seems a bit cumbersome to me here... 😉

**Review by TVolden — APPROVED** [review]

**Comment by TVolden** [comment]
Hi @m-oben,

Everything looked fine, I'm merging it with master.

Thank you for this awesome contribution. I really appreciate it.

Sincerely,
Thomas Volden
